Use decimal precision for float slider

maxsitt opened the pull request:
Fix slider precision mismatch by syncing decimals between line edit and slider steps. Change the default decimal precision from 4 to 2 in FloatSlider, FloatParameter, and MultiFloatParameter.

This makes the float sliders much more precise for use cases with more decimal places. By changing the default decimal precision to 2 for the float parameters, overly precise sliders are avoided if decimals are not set.

maxsitt committed:
Use decimal precision for float slider

Fix slider precision mismatch by syncing decimals between line edit and slider steps. Change the default decimal precision from 4 to 2 in FloatSlider, FloatParameter, and MultiFloatParameter.

maxsitt committed:
Improve slider step and tick interval logic

Refine the calculation of page step and tick interval for IntSlider and FloatSlider based on range size and decimal precision, providing more intuitive slider behavior. Ensure that FloatParameter and MultiFloatParameter sliders update their decimal precision when set.

beatreichenbach commented:
Thank you so much for your PRs, I will take a look at them this week!

maxsitt commented:
Hi Beat,

thanks for your amazing package! I also love your other qt packages, but just getting started with them.

I hope my changes make sense from your design perspective, as more ticks and finer intervals also potentially mean more visual clutter. I tested it with many different ranges and decimal places and for my use case it works better now.

Let me know if there is anything I can contribute further!

beatreichenbach committed:
style: format docstrings

beatreichenbach commented:
Hi Max,
I formatted your PR, making sure there's a space after the docstring and keeping black line length.

> thanks for your amazing package! I also love your other qt packages, but just getting started with them.

Thank you! I'm happy you can find use of them!

> I hope my changes make sense from your design perspective, as more ticks and finer intervals also potentially mean more visual clutter. I tested it with many different ranges and decimal places and for my use case it works better now.

Hm, can you share more about your use case? If I look at the `tests/widgets.py` file which I use to test the widgets, I prefer the current ticks better. Please let me know what range you use.
<img width="408" height="445" alt="sliders_proposed" src="https://github.com/user-attachments/assets/80c4f657-1baf-4457-99e7-638ad16062bb" />

Here's the proposed changes which feel way too dense for me, no?
<img width="407" height="445" alt="sliders_current" src="https://github.com/user-attachments/assets/91fb5e1e-e879-4821-9234-e6dce3f59dae" />

maxsitt committed:
Limit tick count and use nice intervals

Use a 'nice number' algorithm for tick and step intervals, resulting in more human-friendly UI behavior for the InSlider and FloatSlider. Set the target tick count to 10. Add latest changes regarding locale for validator and string conversion of text.

maxsitt commented:
Looking at it again I think your absolutely right - enabling a maximum possible number of ~100 ticks was way too much. I changed the maximum tick count to ~10 and added a `_nice_number` function to calculate more human-friendly tick intervals in  https://github.com/beatreichenbach/qt-parameters/pull/3/commits/b7ba3ebee78bcde268f233b50679d7a6fe830aaa

The main improvement is still the ability to use the `FloatSlider` with decimal precision, which is important for my use case regarding UX.

Can you check if this works for you and do you think these changes make sense?

beatreichenbach commented:
Hi,
I added a test file https://github.com/beatreichenbach/qt-parameters/blob/main/tests/number.py

I don't see what the problem is with the current ticks. Please provide screenshots or test examples, otherwise I can't understand what your issue is.

> This makes the float sliders much more precise for use cases with more decimal places. 

The test widgets work with 4 decimals the same as they work with 1 decimals. 

<img width="1279" height="562" alt="image" src="https://github.com/user-attachments/assets/17d98c8a-5c94-45f4-8226-7ff5836bda97" />

maxsitt commented:
Hi Beat,

thanks for adding the `number.py` test file! This helps a lot and sorry for my previous assumptions that might have been a little bit confusing.

Here are some general points that don't seem to work as expected, or I might misunderstand something:

1. When moving the Int sliders to >= 1000, the value in the line editor resets to 0. This does not happen for the Float sliders.
2. The step for the `Int 1K` slider is 1 (which is good), but the `Float 1K` slider has a step of 10.
3. There are only two ticks at the first and last value for the Int slider with a maximum value of < 20. For the `Float 10` slider there is a tick at every round number (which is good).

---

Regarding the decimal precision of the float sliders:

When trying to use 2 decimals for the `Float 10` slider, the following code leads to the slider only giving 1 decimal precision:
```
parm = FloatParameter('float_10')
parm.set_decimals(2)
parm.set_slider_min(0.00)
parm.set_slider_max(10.00)
parm.set_default(1.00)
```
Changing the default to `1.01` leads to the same result. Changing the minimum slider value to `0.01` gives correct decimal precision for values < 5, but jumps back to only one decimal for values > 5. The opposite happens when changing the maximum slider value to `10.01`, then the correct decimal precision of 2 starts for values > 5. Only changing both the minimum slider value to `0.01` and the maximum slider value to `10.01` gives correct decimal precision for the whole range. Maybe this is caused by some rounding operations in the background? Also the slider step in this case is `0.1` instead of the expected `0.01`.

A similar thing happens for the `Float 0.1` slider where the precision is actually set to 4 decimals, but the slider only gives a precision of 3 decimals. Changing `slider_min` to `0.0001` and `slider_max` to `0.1001` gives the expected precision, but again with a slider step of `0.001` instead of the expected `0.0001`.

Maybe everything actually works as it should and I'm just misunderstanding something or have wrong expectations. Any clarifications would be much appreciated! Also please let me know if I should add a gif that shows this behaviour, but I think it is easy to reproduce.

beatreichenbach commented:
Hi @maxsitt 
Yes I think there's a misunderstanding. The difference between the QSlider and the QLineEdit values are expected. The QSlider has a different "accuracy" than the QLineEdit. The QLineEdit has decimals while the QSlider has a step size.

I committed a change to make this step size dynamic with the addition of a step factor. The higher this factor, the more fine grained the step size of the slider is. https://github.com/beatreichenbach/qt-parameters/commit/de50f6d6df4ce2acea96e223baf85344453ec463

You can see the example in [tests/number.py](https://github.com/beatreichenbach/qt-parameters/blob/main/tests/number.py).

I consider this issue closed. Feel free to open a new issue or continue here if it's relevant.
